### PR TITLE
[Bulky] Admin page for editing bulky items list

### PR DESCRIPTION
**Pull request by davea**
Adds a new page to the admin that displays items in a table:

image

This is linked at the bottom of the `/admin/waste/2566` page:
image

You can edit items, add new ones, or delete existing items, and it all gets stored in the config JSON:

https://user-images.githubusercontent.com/4776/193123593-a75ae9ed-35d7-4393-b59d-e67e40597f28.mov

I reused the existing JS that dynamically adds/removes rows, and figured out how to get the header to stick in place when scrolling a really long table :)

@nephila-nacrea let me know if the fields/data structure isn't what you expected for the part you're working on.

[skip changelog]

For https://github.com/mysociety/societyworks/issues/2914

### CODE REVIEW

File: t/app/controller/waste_peterborough.t [hunk 1]
```diff
@@ -821,9 +822,137 @@ FixMyStreet::override_config {
...
+
+        subtest 'Bartec feature list is shown correctly' => sub {
+            $body->set_extra_metadata(wasteworks_config => {});
+            $body->update;
+
+            $b->mock('Features_Types_Get', sub { [
+                {
+                    Name => "Bookcase",
+                    ID => 6941,
+                    FeatureClass => {
```
Comment (nephila-nacrea) [inline t/app/controller/waste_peterborough.t:900]: Is FeatureClass something like a category? Is there a way to convert the ID to a category string?

**Review by nephila-nacrea — CHANGES_REQUESTED** [review]

File: t/app/controller/waste_peterborough.t (hunk 1 above)
Comment (davea) [reply t/app/controller/waste_peterborough.t:900]: Not really - it's either `BULKY` or `BULKY WHITES`, I believe.

File: templates/web/base/admin/waste/bulky_items.html [hunk 2]
```diff
@@ -0,0 +1,68 @@
+[% INCLUDE 'admin/header.html' title=loc('Bulky items list') -%]
+[% PROCESS 'admin/report_blocks.html' %]
+
+[% INCLUDE status_message %]
+
+[% BLOCK item_row %]
+    <tr class="js-metadata-item">
+        <td>
+            [% IF available_features.size %]
+                <select class="form-control" name="bartec_id[[% i %]]"[% IF i < 9999 %] required[% ELSE %] data-required[% END %]>
```
[the remark this answers is not in the document]
Comment (davea) [reply templates/web/base/admin/waste/bulky_items.html:10]: It's a magic number that the JS looks for, don't know if there's much point variable-ising it.

File: perllib/FixMyStreet/App/Controller/Admin/Waste.pm [hunk 3]
```diff
@@ -66,7 +83,46 @@ sub edit : Path : Args(1) {
...
+        foreach my $i (@indices) {
+            my $bartec_id = $c->get_param("bartec_id[$i]");
+            next unless $bartec_id;
+            push @items, {
+                bartec_id => $c->get_param("bartec_id[$i]"),
+                category => $c->get_param("category[$i]"),
+                name => $c->get_param("name[$i]"),
+                message => $c->get_param("message[$i]"),
+                price => $c->get_param("price[$i]"),
+            };
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/App/Controller/Admin/Waste.pm:121]: That's a bit irritating but I'm not too concerned as the raw JSON editor is only a temporary thing while we build the admin UI to manage that data.

File: perllib/FixMyStreet/App/Controller/Admin/Waste.pm [hunk 4]
```diff
@@ -66,7 +83,46 @@ sub edit : Path : Args(1) {
...
+        }
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/App/Controller/Admin/Waste.pm:150]: Validation is now in place :)

File: perllib/FixMyStreet/App/Controller/Admin/Waste.pm [hunk 5]
```diff
@@ -106,35 +106,48 @@ sub bulky_items : Chained('body') {
...
                 price => $c->get_param("price[$i]"),
             };
+            # OK to store errors in $item itself as it won't get persisted,
+            # and $i might not be the same when form is re-rendered.
+            foreach (qw(name category bartec_id)) {
+                if (!$item->{$_}) {
+                    $item->{errors} ||= {};
+                    $item->{errors}->{$_} = _("This field is required.");
+                    $c->stash->{errors} = 1;
+                }
```
Comment (nephila-nacrea) [inline perllib/FixMyStreet/App/Controller/Admin/Waste.pm:129]: You may also want to limit the length of inputs such as 'message'; I can currently input vast swathes of text (not that it's likely for someone to do that in reality).

File: perllib/FixMyStreet/App/Controller/Admin/Waste.pm [hunk 6]
```diff
@@ -106,35 +106,48 @@ sub bulky_items : Chained('body') {
...
+            }
+            push @items, $item;
+        }
+        unless ($c->stash->{errors}) {
+            # sort by category and then name
+            @items = sort {
```
Comment (nephila-nacrea) [inline perllib/FixMyStreet/App/Controller/Admin/Waste.pm:135]: The items are being sorted here, but come out in random order on the admin page.

**Review by nephila-nacrea — CHANGES_REQUESTED** [review]

File: perllib/FixMyStreet/App/Controller/Admin/Waste.pm [hunk 7]
```diff
@@ -106,35 +106,48 @@ sub bulky_items : Chained('body') {
...
                 category => $c->get_param("category[$i]"),
                 name => $c->get_param("name[$i]"),
                 message => $c->get_param("message[$i]"),
                 price => $c->get_param("price[$i]"),
             };
+            # OK to store errors in $item itself as it won't get persisted,
+            # and $i might not be the same when form is re-rendered.
+            foreach (qw(name category bartec_id)) {
+                if (!$item->{$_}) {
+                    $item->{errors} ||= {};
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/App/Controller/Admin/Waste.pm:126]: Ah, I didn't know Perl would do that. I'll leave it in to make it clearer.

File: perllib/FixMyStreet/App/Controller/Admin/Waste.pm (hunk 6 above)
Comment (davea) [reply perllib/FixMyStreet/App/Controller/Admin/Waste.pm]: Items now always retain the same order, it doesn't do any sorting by name at all.

File: templates/web/base/admin/waste/bulky_items.html [hunk 8]
```diff
@@ -4,7 +4,7 @@
 [% INCLUDE status_message %]
 
 [% BLOCK item_row %]
```
[the remark this answers is not in the document]
Comment (davea) [reply templates/web/base/admin/waste/bulky_items.html:6]: Ends up as a bit of a mess! I've improved the validation to stop that from happening, and also added the delete button for non-JS users.

**Review by nephila-nacrea — APPROVED** [review]
Only thing that occurred to me is whether we want 'price' to be a numeric field, but I am not sure if there is a use case where the user wants to use that field for a message as well.

**Pull request merged**